tests/components/google_air_quality/test_config_flow.py

```diff
@@ -0,0 +1,399 @@
+    await hass.async_block_till_done()
+    assert config_entry.state is ConfigEntryState.LOADED
+
+    result = await hass.config_entries.subentries.async_init(
+        (config_entry.entry_id, "location"),
+        context={"source": SOURCE_USER},
+    )
+    assert result["type"] is FlowResultType.FORM
+    assert result["step_id"] == SOURCE_USER
+
+    result = await hass.config_entries.subentries.async_configure(
+        result["flow_id"],
+        user_input={
+            CONF_LOCATION: {
+                CONF_LATITUDE: 0,
+                CONF_LONGITUDE: 0,
+            }
+        },
+    )
+    assert result["type"] is FlowResultType.ABORT
+    assert result["reason"] == reason
+
+
+@pytest.mark.usefixtures(
+    "current_request_with_host",
+)
+@pytest.mark.parametrize("updated_token_entry", [{"scope": "nada"}])
+async def test_missing_scope(
+    hass: HomeAssistant,
+    hass_client_no_auth: ClientSessionGenerator,
+    mock_setup: Mock,
+) -> None:
+    """Check if the scopes are there, before prestenting the map."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+    state = config_entry_oauth2_flow._encode_jwt(
+        hass,
+        {
+            "flow_id": result["flow_id"],
+            "redirect_uri": "https://example.com/auth/external/callback",
+        },
+    )
+
+    client = await hass_client_no_auth()
+    resp = await client.get(f"/auth/external/callback?code=abcd&state={state}")
+    assert resp.status == 200
+
+    step1 = await hass.config_entries.flow.async_configure(result["flow_id"])
+    assert step1["type"] is FlowResultType.ABORT
+
+
+@pytest.mark.usefixtures("setup_integration")
+@pytest.mark.usefixtures("current_request_with_host", "mock_api")
+@pytest.mark.parametrize(
+    ("fixture_name", "api_error"),
+    [("air_quality_data.json", NoDataForLocationError())],
+)
+async def test_no_data_for_location_shows_form(
+    hass: HomeAssistant,
+    config_entry: MockConfigEntry,
+    hass_client_no_auth: ClientSessionGenerator,
+    mock_setup: Mock,
+) -> None:
+    """Show form with base error if no data is available for the location."""
+
+    await hass.async_block_till_done()
+    assert config_entry.state is ConfigEntryState.LOADED
+
+    result = await hass.config_entries.subentries.async_init(
+        (config_entry.entry_id, "location"),
+        context={"source": SOURCE_USER},
+    )
+    assert result["type"] is FlowResultType.FORM
+    assert result["step_id"] == SOURCE_USER
+
+    step2 = await hass.config_entries.subentries.async_configure(
+        result["flow_id"],
+        user_input={
+            CONF_LOCATION: {
+                CONF_LATITUDE: 0,
+                CONF_LONGITUDE: 0,
+            }
+        },
+    )
+    assert step2["type"] is FlowResultType.FORM
+    assert step2["step_id"] == SOURCE_USER
+    assert step2["errors"] == {"base": "no_data_for_location"}
```

Comment:
can we finish this flow?

Reply:
done in https://github.com/home-assistant/core/pull/145237/commits/1b2707eda7755ea6af6da23f1a0ca329c124b6c0